### ISSUE TITLE
743. Network Delay Time

### ISSUE THREAD

**Issue by songyy5517** [songyy5517/DataStructures-Algorithms#123]
You are given a network of n nodes, labeled from 1 to n. You are also given times, a list of travel times as directed edges times[i] = (ui, vi, wi), where ui is the source node, vi is the target node, and wi is the time it takes for a signal to travel from source to target.
We will send a signal from a given node k. Return the minimum time it takes for all the n nodes to receive the signal. If it is impossible for all the n nodes to receive the signal, return -1.

Example 1:
![image](https://github.com/user-attachments/assets/98f399a7-a853-4bfa-9f12-b6a4afd3677c)
```
Input: times = [[2,1,1],[2,3,1],[3,4,1]], n = 4, k = 2
Output: 2
```

**Intuition**
Basically, we need to find the shortest path from the source node to all other nodes, and return the maximum value of path. Given that the weights in the problem are always positive, therefore, we can employ Dijkstrua to solve this problem.

**Comment by songyy5517** [comment]
**Approach: Dijkstrua's Algorithm**
1. Exception handling;
2. Define variables & Initialization:
- `class iPair`: stores the neighbor of a node and their weight;
- `int[] dist`: records the value of the shortest path from the source node to other nodes;
- `PriorityQueue<iPair> min_Heap`: used to pick the node with the minimum path every time;
- `ArrayList<ArrayList<iPair>> graph`: stores the graph in a adjacent list;
3. Put the source node in the priorityqueue;
4. While the priorityqueue is not empty:
（1）Obtain the node from the queue with the minimum path;
（2）Check all its neighbors, if the current path value is less than their previous one, then update, and add these neighbor pairs into the queue.
5. Find the maximum value in the dist array, return it if it is valid; otherwise, return -1.

**Complexity Analysis**
- Time complexity: $O(V\log V + E\log V)$ ，in 4(1), all nodes in the graph need to be added and removed in the min_Heap, which costs $O(V\log V)$ time; in 4(2), all the neighbors of each node in the graph are added in the min_Heap, that is, the add operation $O(\log V)$ need to be performed E times, totally taking up $O(E\log V)$ time.
- Space complexity: $O(V+E)$ , the dist array and the priorityqueue occupy $O(V)$ space, and the adjacent list storing the whole graph takes up $O(V+E)$ space.

**Comment by songyy5517** [comment]
```
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        // Dijkstrua
        // 1. Exception handling
        if (times == null || times.length == 0 || n < 1 || k < 1 || k > n)
            return -1;
        
        // 2. Create an adjacent list for the graph
        ArrayList<NodeDist>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++)
            graph[i] = new ArrayList();
        for (int i = 0; i < times.length; i++){
            int u = times[i][0], v = times[i][1], w = times[i][2];
            graph[u - 1].add(new NodeDist(v - 1, w));
        }

        // 3. Define variables and initialize
        int[] dist = new int[n];    // store the shortest distance from the source node to each other node
        // min_heap: stores unvisited nodes. At every movement, we extract the node with the minimum distance
        // and finalize, which means that the node is moved from unvisted to visited.
        PriorityQueue<NodeDist> min_heap = new PriorityQueue<NodeDist>((x, y) -> x.weight - y.weight);

        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[k - 1] = 0;
        min_heap.add(new NodeDist(k - 1, 0));

        // 4. Calculate the shortest distance of all nodes
        while (!min_heap.isEmpty()){
            // Pick a unvisited node with the minimum distance in the min_heap
            NodeDist cur_node = min_heap.remove();
            int node = cur_node.node;

            // Relax: Check its neighbors, and update their distance if the current path is shorter.
            for (int i = 0; i < graph[node].size(); i++){
                int neighbor = graph[node].get(i).node, weight = graph[node].get(i).weight;
                int cur_dist = dist[node] + weight;

                if (cur_dist < dist[neighbor]){
                    dist[neighbor] = cur_dist;
                    // If the neighbor's distance changes, then the distance of its neighbors also might change, 
                    // so we need to put it into the min_heap. Thus, we can update its neighbors' distance.
                    min_heap.add(new NodeDist(neighbor, dist[neighbor]));    
                }
            }
        }

        // 5. Find the maximum distance of all nodes
        int max = 0;
        for (int i : dist)
            max = Math.max(max, i);
        
        return max == Integer.MAX_VALUE ? -1 : max;
    }
}

class NodeDist {
    int node;
    int weight;

    NodeDist(int n, int w){
        node = n;
        weight = w;
    }
}
```

**Comment by songyy5517** [comment]
2024/11/13
2024/12/2
1. Why is the github code faster?
- Convert the graph to an adjacent list. This way, for each node in the minimum heap, we can just check its neighbor edges, instead of all edges in the graph, which saves a lot of time.
2. Complexity Analysis
- Time complexity of `PriorityQueue min_heap`
3. Algorithm prodecures of Dijkstrua.
4. PriorityQueue中lambda表达式前必须加上类型声明.
5. 为什么用PriorityQueue而不是Queue?
  模拟每次选取路径值最小的节点。如果用队列，每次选取的节点不一定是当前的最短路径，所以此时针对该节点的finalize操作是无效的。
